internal/http_client.go

```diff
@@ -0,0 +1,193 @@
+}
+
+// Do executes the given Request, and returns a Response.
+func (c *HTTPClient) Do(ctx context.Context, r *Request) (*Response, error) {
+	req, err := r.newHTTPRequest()
+	if err != nil {
+		return nil, err
+	}
+
+	resp, err := c.Client.Do(req.WithContext(ctx))
+	if err != nil {
+		return nil, err
+	}
+	defer resp.Body.Close()
+
+	b, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		return nil, err
+	}
+	return &Response{
+		Status:    resp.StatusCode,
+		Body:      b,
+		Header:    resp.Header,
+		errParser: c.ErrParser,
+	}, nil
+}
+
+// Request contains all the parameters required to construct an outgoing HTTP request.
+type Request struct {
+	Method string
+	URL    string
+	Body   HTTPEntity
+	Opts   []HTTPOption
+}
+
+func (r *Request) newHTTPRequest() (*http.Request, error) {
+	var opts []HTTPOption
+	var data io.Reader
+	if r.Body != nil {
+		b, err := r.Body.Bytes()
+		if err != nil {
+			return nil, err
+		}
+		data = bytes.NewBuffer(b)
+		opts = append(opts, WithHeader("Content-Type", r.Body.Mime()))
+	}
+
+	req, err := http.NewRequest(r.Method, r.URL, data)
+	if err != nil {
+		return nil, err
+	}
+
+	opts = append(opts, r.Opts...)
+	for _, o := range opts {
+		o(req)
+	}
+	return req, nil
+}
+
+// HTTPEntity represents a payload that can be included in an outgoing HTTP request.
+type HTTPEntity interface {
+	Bytes() ([]byte, error)
+	Mime() string
+}
+
+type jsonEntity struct {
+	Val interface{}
+}
+
+// NewJSONEntity creates a new HTTPEntity that will be serialized into JSON.
+func NewJSONEntity(v interface{}) HTTPEntity {
+	return &jsonEntity{Val: v}
+}
+
+func (e *jsonEntity) Bytes() ([]byte, error) {
+	return json.Marshal(e.Val)
+}
+
+func (e *jsonEntity) Mime() string {
+	return "application/json"
+}
+
+// Response contains information extracted from an HTTP response.
+type Response struct {
+	Status    int
+	Header    http.Header
+	Body      []byte
+	errParser ErrorParser
+}
+
+// CheckStatus checks whether the Response status code has the given HTTP status code.
+//
+// Returns an error if the status code does not match. If an ErroParser is specified, uses that to
+// construct the returned error message. Otherwise includes the full response body in the error.
+func (r *Response) CheckStatus(want int) error {
+	if r.Status == want {
+		return nil
+	}
+
+	var msg string
+	if r.errParser != nil {
+		msg = r.errParser(r.Body)
+	}
+	if msg == "" {
+		msg = string(r.Body)
+	}
+	return fmt.Errorf("http error status: %d; reason: %s", r.Status, msg)
+}
+
+// Unmarshal checks if the Response has the given HTTP status code, and if so unmarshals the
+// response body into the variable pointed by v.
+//
+// Unmarshal uses https://golang.org/pkg/encoding/json/#Unmarshal internally, and hence v has the
+// same requirements as the json package.
+func (r *Response) Unmarshal(want int, v interface{}) error {
+	if err := r.CheckStatus(want); err != nil {
+		return err
+	} else if err := json.Unmarshal(r.Body, v); err != nil {
```

Comment:
I would probably leave off the redundant "else", since there's a return in there.

Reply:
Done